File: src/models/pharmacodynamics.ts
/**
 * Pharmacodynamics — Local Anti-Inflammatory Efficacy
 * Computes therapeutic effect using the classical Emax Hill equation:
 * E(Dt) = (Emax * Dt) / (EC50 + Dt)
 */
export function calculateEfficacy(dt: number, emax: number, ec50: number): number {
  if (dt <= 0 || ec50 <= 0) return 0;
  return (emax * dt) / (ec50 + dt);
}

/**
 * Calculates trapezoidal Area Under the Curve (AUC) across an array of (time, value) pairs.
 */
export function calculateAUC(points: Array<{ time: number; val: number }>): number {
  let auc = 0;
  for (let i = 0; i < points.length - 1; i++) {
    const dt = points[i + 1].time - points[i].time;
    const avgVal = (points[i].val + points[i + 1].val) / 2;
    auc += avgVal * dt;
  }
  return auc;
}

File: models/pharmacodynamics.py
"""
Pharmacodynamics — Local Pulmonary Anti-Inflammatory Efficacy
Emax Hill Model:
  E(Dt) = (Emax * Dt) / (EC50 + Dt)
"""

import numpy as np

def calculate_efficacy(dt_array: np.ndarray, emax: float = 100.0, ec50: float = 0.5) -> np.ndarray:
    """
    Computes anti-inflammatory efficacy percentage as a function of target tissue concentration Dt.
    
    Parameters:
        dt_array (np.ndarray or float): Target tissue concentration or amount Dt (mg).
        emax (float): Maximum achievable therapeutic effect (%) [default 100%].
        ec50 (float): Tissue concentration producing 50% of Emax (mg).
        
    Returns:
        np.ndarray: Anti-inflammatory efficacy E (%)
    """
    dt_clean = np.maximum(0.0, np.asarray(dt_array, dtype=float))
    denominator = ec50 + dt_clean
    # Avoid zero division
    denominator = np.where(denominator <= 0.0, 1e-9, denominator)
    return (emax * dt_clean) / denominator

def calculate_auc(time_grid: np.ndarray, values: np.ndarray) -> float:
    """
    Computes trapezoidal Area Under the Curve (AUC).
    """
    return float(np.trapz(values, time_grid))
